subprojects/docs/src/docs/userguide/upgrade_version_4.adoc

```diff
@@ -0,0 +1,493 @@
+ * link:{groovyDslPath}/org.gradle.nativeplatform.tasks.AbstractLinkTask.html#org.gradle.nativeplatform.tasks.AbstractLinkTask:linkerArgs[AbstractLinkTask.linkerArgs] is now of type `ListProperty<String>` instead of `List<String>`.
+ * `TaskDestroyables.getFiles()` is no longer part of the public API.
+ * Overlapping version ranges for a dependency now result in Gradle picking a version that satisfies all declared ranges.
++
+For example, if a dependency on `some-module` is found with a version range of `[3,6]` and also transitively with a range of `[4,8]`, Gradle now selects version 6 instead of 8. The prior behavior was to select 8.
+ * The order of elements in `Iterable` properties marked with either `@OutputFiles` or `@OutputDirectories` now matters. If the order changes, the property is no longer considered up to date.
++
+Prefer using separate properties with `@OutputFile`/`@OutputDirectory` annotations or use `Map` properties with `@OutputFiles`/`@OutputDirectories` instead.
+ * Gradle will no longer ignore dependency resolution errors from a repository when there is another repository it can check. Dependency resolution will fail instead. This results in more deterministic behavior with respect to resolution results.
+
+[[upgrade_4.1]]
+== From 4.1
+
+=== Deprecations
+
+ * You should no longer use any of the following characters in domain object names, such as project and task names: <space> `/ \ : < > " ? * |`. You should also not use `.` as a leading or trailing character.
+
+=== Potential breaking changes
+
+ * The `withPathSensitivity()` methods on link:{javadocPath}/org/gradle/api/tasks/TaskFilePropertyBuilder.html[TaskFilePropertyBuilder] and link:{javadocPath}/org/gradle/api/tasks/TaskOutputFilePropertyBuilder.html[TaskOutputFilePropertyBuilder] have been removed.
+ * The bundled `bndlib` has been upgraded from 3.2.0 to 3.4.0.
+ * The FindBugs Plugin no longer renders progress information from its analysis. If you rely on that output in any way, you can enable it with link:{groovyDslPath}/org.gradle.api.plugins.quality.FindBugs.html#org.gradle.api.plugins.quality.FindBugs:showProgress[FindBugs.showProgress].
+
+[[upgrade_4.0]]
+== From 4.0
+
+ * Consider using the new <<custom_tasks.adoc#worker_api,Worker API>> to enable units of work within your build to run in parallel.
+
+=== Deprecated classes, methods and properties
+
+Follow the API links to learn how to deal with these deprecations (if no extra information is provided here):
+
+ * link:{javadocPath}/org/gradle/api/tasks/scala/ScalaDocOptions.html#getStyleSheet--[ScalaDocOptions.styleSheet] — the Scaladoc Ant task in Scala 2.11.8 and later no longer support this property.
+ * link:{javadocPath}/org/gradle/api/Nullable.html[Nullable]
+ * link:{javadocPath}/org/gradle/api/Task.html#dependsOnTaskDidWork--[Task.dependsOnTaskDidWork()]
```

Comment:
Should we say that there's no substitute?

Reply:
We should say something, yes. I'll review the release notes again and ask others if necessary.

Reply:
Actually, isn't the Javadoc sufficient here? It explains that users should use declared inputs and outputs instead.